RavenDB-15998 handle cluster tx conflicts and test fixes

karmeli87 opened the pull request.

karmeli87 committed:
RavenDB-15998 properly handle conflicts of cluster tx

- upon conflict between normal document and a remote cluster tx document
the remote cluster tx document should preserve cluster tx flags

- conflict between two documents that were modified in _different_ cluster, should be treated as a normal conflict.

karmeli87 committed:
RavenDB-15998 tight MissingRevisions tests

ppekrol commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -1773,6 +1776,37 @@ public void EnsureLastEtagIsPersisted(DocumentsOperationContext context, long do
             return (newEtag, changeVector);
         }
 
+        public static bool FlagsProperlySet(DocumentFlags flags, string changeVector, out string reason)
```
I'm worried that in DEBUG we will be doing that. What is the slowdown?

ayende replied:
Yes, we want that with `Conditional`, I think

ppekrol replied:
Sure, but we are using this in Debug only, I'm worried that this can get us severe slowdown during Debug tests, but maybe I'm not right

karmeli87 replied:
I don't think there is a severe slowdown here... 
I guess the most consuming part is parsing the change-vector and most of our tests has a change-vector of 1 or 2 entries.

ppekrol commented on src/Raven.Server/Documents/DocumentsStorage.cs:
```diff
@@ -1773,6 +1776,37 @@ public void EnsureLastEtagIsPersisted(DocumentsOperationContext context, long do
+                    reason = $"FromClusterTransaction, expect change vector of length 1, {changeVector}";
+                    return false;
+                }
+                if (cvArray[0].NodeTag != "RAFT".ParseNodeTag())
```
`RAFT` looks like a constant. Don't we have a place for those?

ppekrol replied:
Is `ParseNodeTag` for each call necessary?

ayende commented on src/Raven.Server/Documents/Replication/IncomingReplicationHandler.cs:
```diff
@@ -1418,6 +1418,18 @@ protected override long ExecuteCmd(DocumentsOperationContext context)
                                             _replicationInfo.Logger.Info(
                                                 $"Conflict check resolved to Conflict operation, resolving conflict for doc = {item.Id}, with change vector = {item.ChangeVector}");
 
+                                        if (hasLocalClusterTx == hasRemoteClusterTx)
```
Can this be that both are false here?

karmeli87 replied:
yes, both false is the common case, normal conflict

karmeli87 committed:
RavenDB-15998 move FlagsProperlySet to conditional debug

karmeli87 committed:
RavenDB-15998 remove ResolveInFavorOfLocalClusterTransaction test

The issue is that if we set a master-master replication between two clusters
we might end up in a situation where we have 2 different documents with the same change vector,
since each cluster will resolve it to his local document and merge the change vectors

karmeli87 committed:
RavenDB-15998 don't change local document if it from cluster tx
